SC-103: Complete all testing for the FundingManager

Zitzak opened the pull request:
## What has been done?
- Completed all the testing
- Moved function from implementation into base contracts
- Added functions to base contract
- Made all functions virtual in base contracts

## Implements
- https://linear.app/inverter/issue/SC-103/complete-all-testing-for-the-funding-manager

Zitzak committed:
fix: update spotPrice function to work with params

The spotPrice function had to be fixed after adapting it
to work with parameters instead of state values

Zitzak committed:
test(ToposFormula.t.sol): add test for Topos Formula

This commit includes the ToposFormulaMock contract

Zitzak committed:
feat: update bonding curve base contracts

- Made all function virtual
- Add functions to get fee amount base on a sale or purchase amount
- Create function to get fee collected
- Move functions to calculate sale and purchase return into base contracts
- Update testing

Zitzak committed:
feat: add check for Topos Formula and add tests

- Moved error check to the tokenOut function
- Added tests

Zitzak committed:
feat: complete ToposBondingCurveFundingManager

- Add init parameters
- Add internal functions
- Add tests

Zitzak committed:
Merge branch 'feature/ToposFundingManager' into sc-103-complete-all-testing-for-the-funding-manager

Zitzak commented on test/modules/fundingManager/bondingCurveFundingManager/ToposBondingCurveFundingManager.t.sol:
```diff
@@ -127,10 +183,1571 @@ contract ToposBondingCurveFundingManagerTest is ModuleTest {
+            .call_redeemTokensFormulaWrapper(_depositAmount);
+
+        // Because of precision loss, the assert is done to be in range of 0.0000000001% of each other
+        assertApproxEqRel(functionReturnValue, redeemAmount, 0.0000000001e18);
```
@marvinkruse I could not get both values equal, which seemed odd to me, as the precision loss should be the same if there is any. Would you know why? The diff is very minimal which I resolved in the testing like this

marvinkruse replied:
Hm.. weird indeed, I expected it to be the same as well. DId you check what the difference is when it appears? Is it just 1 wei or anything bigger than that? Then at least it might be apparent that it's a clear precision loss from the division in for example the `basePriceToCapitalRatio`. But not sure why it would lead to different results, both is done in Solidity with the same values. Or is the order in which these calcs happen different?

I will check again in the morning after a coffee, maybe I see it then..  In any case, the solution you have here is good enough for me 😁

Zitzak replied:
So the difference was not a constant but changed with the fuzzing, and as I tried to define the allowed delta. So the value I now came up with works, but it's indeed weird because basically the same function with the same values get called

marvinkruse replied:
Hm.. weird indeed. I don't think this should block the PR though, even if I'd like to understand this eventually. Let's go with the solution you have here and maybe either get the auditors to verify this or revisit this on our own down the road (before we deploy to mainnet).

marvinkruse commented on src/modules/fundingManager/bondingCurveFundingManager/ToposBondingCurveFundingManager.sol:
```diff
@@ -132,13 +134,21 @@ contract ToposBondingCurveFundingManager is
         _setCapitalRequired(bondingCurveProperties.capitalRequired);
+        // Set current seize to MAX_SEIZE
+        _setSeize(bondingCurveProperties.seize);
+        // Set initial
```
Should this say `/// Set initial base price multiplier` or is this intentional?

Zitzak replied:
You're right, did not finish the sentence. Will update

marvinkruse replied:
Thank you! If I wanna be in turbo-pedantic mode I would say that it should say "expressed" instead of "expresses", but you can also ignore this :D Thanks for changing the sentence!

marvinkruse commented on src/modules/fundingManager/bondingCurveFundingManager/ToposBondingCurveFundingManager.sol:
```diff
@@ -132,13 +134,21 @@ contract ToposBondingCurveFundingManager is
-        _setSellFee(MAX_SELL_FEE);
+        _setSellFee(bondingCurveProperties.sellFee);
+        // Set buying functionality to open if true. By default buying is false
+        if (bondingCurveProperties.buyIsOpen == true) _openBuy();
+        // Set selling functionality to open if true. By default selling is false
+        if (bondingCurveProperties.sellIsOpen == true) _openSell();
```
Just caught my eye that you're using `bondingCurveProperties.buyIsOpen == true` instead of just `bondingCurveProperties.buyIsOpen` which does the same. Do you just do this for simplicity in terms of reading? Was just curious as I think the Bonding Curve contracts are the only ones that do this afaik. Not important and doesn't necessarily need a change, just out of curiosity. It compiles to the same code at least so no difference functionally.

Zitzak replied:
Yeah I tend to do this for clarity. I do always try to opt for readability as you've prob have seen in the tests as well :))

However a question back to you would be if there are any negative gas associations to this, in which case I would def change it

marvinkruse replied:
No gas implications, as it both compiles to the assembly code, so all good. Ideally, I'd love to have the codebase be consistent, which it currently isn't, but let's just keep this like it is and discuss this at a different time with the rest briefly to decide on a general level, I'd say.

marvinkruse commented on src/modules/fundingManager/bondingCurveFundingManager/ToposBondingCurveFundingManager.sol:
```diff
@@ -457,12 +452,22 @@ contract ToposBondingCurveFundingManager is
+            revert ToposBondingCurveFundingManager__InvalidSeize(_seize);
+        }
+        currentSeize = _seize;
+    }
```
Do we want an event for updating the `currentSeize`? Just thinking in terms of our goal that we would ideally like to be able to rebuild the entire onchains state from events for potential indexing solutions.

marvinkruse replied:
There is one for example for `CapitalRequiredChanged`, maybe it's more consistent to have one here as well there?

Zitzak replied:
You're right, makes sense. It wasn't in the original contracts weirdly enough so overlooked it

Zitzak replied:
Quick question in regard to "being able to rebuild the on-chain state for indexing". The collateral fee collected value is internal and no event emit reflects this state. My thought process was that it shouldn't be easily retrievable for everyone how much fee has been collected. However this would be one value that is not rebuildable through emits in this way.

marvinkruse replied:
Hm.. good question. I think we don't need an event for changes to `tradeFeeCollected`, seems a bit overkill for me to have one if I think about it.

As it is exposed via `getCollateralTradeFeeCollected()` in `BondingCurveFundingManagerBase `, it is not hidden though, not sure whether you meant that? In the end, an indexer can still track it as it's publically exposed via a this getter function, so we can make use of it in a potential dashboard to track fees etc. (which we probably want at some point). Users are also able to see the value easily - which to me is not a problem, I wouldn't hide it (just not rub it in their face via an event 😁)

Would you agree?

Zitzak replied:
So the `getCollateralTradeFeeCollected()` function has the `onlyOrchestratorOwner` modifier, which would inhibit it from being publicly available.

However reading your comment and seeing that you don't see an issue with users seeing the value, would it then make sense to just remove the getter function and make the `tradeFeeCollected` public, maybe also renaming it so it reflects that collateral fee is collected here?

marvinkruse replied:
Oh, I missed that modifier indeed, good point! I think making the variable itself public is fine, also agree on the renaming. 👍

marvinkruse reviewed (changes requested):
No things on the logic front that I found, all looking solid. Just some minor things again.

Zitzak committed:
chore: address PR comments

- Fixed missing comment
- Added event to function which sets the currentSeize value
- Adapted tests

Zitzak committed:
chore: add events and fix test

- Add events
- Add tests for events
- Fix CI failure

marvinkruse commented on src/modules/fundingManager/bondingCurveFundingManager/ToposBondingCurveFundingManager.sol:
```diff
@@ -132,13 +134,21 @@ contract ToposBondingCurveFundingManager is
         formula = IToposFormula(bondingCurveProperties.formula);
+        // Set capital required
         _setCapitalRequired(bondingCurveProperties.capitalRequired);
+        // Set current seize to MAX_SEIZE
```
Isn't this set to `bondingCurveProperties.seize` instead of `MAX_SEIZE`?

Zitzak replied:
You're right. This is still an old comment. Changed it

Zitzak committed:
fix: rewrite comments for clarity

Zitzak committed:
chore: rename fee collected variable and make it public

marvinkruse reviewed (approved):
great stuff!
